src/software/geom/algorithms/collinear.cpp

```diff
@@ -1,31 +1,23 @@
 #include "software/geom/algorithms/collinear.h"
 
+#include "software/geom/algorithms/acute_angle.h"
 #include "software/geom/algorithms/almost_equal.h"
 
 
-bool collinear(const Point &a, const Point &b, const Point &c, double fixed_epsilon,
-               int ulps_epsilon)
+bool collinear(const Point &a, const Point &b, const Point &c)
 {
-    if ((a - b).lengthSquared() < fixed_epsilon ||
-        (b - c).lengthSquared() < fixed_epsilon ||
-        (a - c).lengthSquared() < fixed_epsilon)
-    {
-        return true;
-    }
-
-    if ((almostEqual(a.x(), b.x(), fixed_epsilon, ulps_epsilon) &&
-         almostEqual(a.x(), c.x(), fixed_epsilon, ulps_epsilon)) ||
-        (almostEqual(a.y(), b.y(), fixed_epsilon, ulps_epsilon) &&
-         almostEqual(a.y(), c.y(), fixed_epsilon, ulps_epsilon)))
-    {
-        // Explicit check for the vectors being near vertical or horizontal to avoid near
-        // zero comparisons
-        return true;
-    }
-
-    Vector v1 = b - a;
-    Vector v2 = c - a;
-    return almostEqual(v1.x() * v2.y(), v1.y() * v2.x(), fixed_epsilon, ulps_epsilon);
+    // NOTE: the default value is 4 * FIXED_EPSILON because of accumulation of error from
+    // subtracting 2 vectors twice and calculating acuteAngle, but the PassGenerator
+    // requires less precision probably due to calcBestShotOnGoal
+    // TODO (#1788): change fixed_epsilon to 4 * FIXED_EPSILON
+    static const double fixed_epsilon = 100 * FIXED_EPSILON;
+    static const int ulps_epsilon     = ULPS_EPSILON_TEN;
+    return (almostEqual(acuteAngle(b - a, c - a).toRadians(), 0, fixed_epsilon,
+                        ulps_epsilon) ||
+            almostEqual(acuteAngle(a - b, c - b).toRadians(), 0, fixed_epsilon,
+                        ulps_epsilon) ||
+            almostEqual(acuteAngle(a - c, b - c).toRadians(), 0, fixed_epsilon,
+                        ulps_epsilon));
```

Comment:
Could you explain what almostEqual() does and what the math here is doing? Sorta confused,

Reply:
the method is trying to find whether 3 points are collinear. To do so, it compares the angle (using `acuteAngle()`) of the segments intersection AB,BC,AC with one another (segment AB is represented by `a-b` in the code). if the points are collinear, meaning there exists a line that goes through them all, then the intersection angle will be 0. 

the reason we use `almostEqual()` is that floating point values arent represented exactly in binary, so comparing them to 0 directly using `==` can sometimes lead to the incorrect result. So we give a small value of `fixed_epsilon` (in this case its `10^(-12)` that gives a maximum difference to what we can consider equal. 

However `fixed_epsilon` doesnt deal with all cases, so we also use `ulps_epsilon`. This [article](https://bitbashing.io/comparing-floats.html) goes into more detail about our implementation of `almostEqual()` and the reasoning behind it.